src/kakarot/instructions/stop_and_math_operations.cairo

```diff
@@ -0,0 +1,565 @@
+        assert opcode = cast(
+            pc + (opcodes_label - pc_label) + opcode_number * model.Opcode.SIZE, model.Opcode*
+        );
+
+        let stack_underflow = is_le(ctx.stack.size, opcode.stack_input - 1);
+        if (stack_underflow != 0) {
+            let (revert_reason_len, revert_reason) = Errors.stackUnderflow();
+            let ctx = ExecutionContext.stop(ctx, revert_reason_len, revert_reason, TRUE);
+            return ctx;
+        }
+        let out_of_gas = is_le(ctx.call_context.gas_limit, ctx.gas_used + opcode.gas - 1);
+        if (out_of_gas != 0) {
+            let (revert_reason_len, revert_reason) = Errors.outOfGas();
+            let ctx = ExecutionContext.stop(ctx, revert_reason_len, revert_reason, TRUE);
+            return ctx;
+        }
+
+        let (local stack, popped) = Stack.pop_n(ctx.stack, opcode.stack_input);
+
+        // offset is 1 (new line) + 2 (jmp + label) per opcode
+        // opcode is offset from by 0x1 (index of the first opcode)
+        tempvar offset = 2 * (opcode_number - 0x01) + 1;
+
+        tempvar range_check_ptr = range_check_ptr;
+        tempvar popped = popped;
+
+        jmp rel offset;
+
+        jmp ADD;  // 0x1
+        jmp MUL;  // 0x2
+        jmp SUB;  // 0x3
+        jmp DIV;  // 0x4
+        jmp SDIV;  // 0x5
+        jmp MOD;  // 0x6
+        jmp SMOD;  // 0x7
+        jmp ADDMOD;  // 0x8
+        jmp MULMOD;  // 0x9
+        jmp EXP;  // 0xa
+        jmp SIGNEXTEND;  // 0xb
+        jmp INVALID;  // 0xc
+        jmp INVALID;  // 0xd
+        jmp INVALID;  // 0xe
+        jmp INVALID;  // 0xf
+        jmp LT;  // 0x10
+        jmp GT;  // 0x11
+        jmp SLT;  // 0x12
+        jmp SGT;  // 0x13
+        jmp EQ;  // 0x14
+        jmp ISZERO;  // 0x15
+        jmp AND;  // 0x16
+        jmp OR;  // 0x17
+        jmp XOR;  // 0x18
+        jmp NOT;  // 0x19
+        jmp BYTE;  // 0x1a
+        jmp SHL;  // 0x1b
+        jmp SHR;  // 0x1c
+        jmp SAR;  // 0x1d
+
+        end:
+        // Parse results
+        // All the jumps share the same return signature, which is bitwise_ptr
+        // and range_check_ptr in implicit args and a Uint256 for the value
+        let bitwise_ptr = cast([ap - 4], BitwiseBuiltin*);
+        let range_check_ptr = [ap - 3];
+        let result = Uint256([ap - 2], [ap - 1]);
+
+        // Rebind args with fp
+        // Function args are in [fp - n - 2: fp - 2]
+        // locals are retrieved from [fp] in the order they are defined
+        let syscall_ptr = cast([fp - 7], felt*);
+        let pedersen_ptr = cast([fp - 6], HashBuiltin*);
+        let ctx = cast([fp - 3], model.ExecutionContext*);
+        let opcode = cast([fp], model.Opcode*);
+        let stack = cast([fp + 1], model.Stack*);
+
+        // Finalize opcode
+        let stack = Stack.push_uint256(stack, result);
+        let ctx = ExecutionContext.update_stack(ctx, stack);
+        let ctx = ExecutionContext.increment_gas_used(ctx, opcode.gas);
+        return ctx;
```

Comment:
should `end` label be at the end of the function, should for lisibility?

Reply:
Well I don't know, I've deferred the actual opcodes below to make sure that the main logic is easy to follow (pre opcode, opcode, post opcode), as it can be harder to understand the register retrieval from pre to post otherwise.